Update error message when automatic gather boostrap fails on cluster fail.

patrickdillon opened the pull request:
Many users seem to misinterpret the failure of the gather command to collect logs as the error message for why their cluster failed to install. An example error message looks like this:

```
ERROR failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
FATAL waiting for Kubernetes API: context deadline exceeded
```
This change will switch the error prefix to INFO. I have opted to hide the error message entirely. I think this is the best approach here, but am definitely open to other ideas.

Here are some examples from internal slack channel where users have been confused:

- August 20: @here Having a problem bringing up aws cluster. Getting ERROR failed to create SSH client, ensure the proper ssh key, I have rebuilt tokens and new are the same as previous.
- July 16: Trying to install OpenShift 4.2 AWS cluster from my laptop. I am getting repeated
````
ERROR failed to create SSH client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain 
FATAL waiting for Kubernetes API: context deadline exceeded 
````
- July 29: Hello, when I try to spin up a 4.2 cluster today for myself, Its failing to create an SSH client - how can I get it to prompt me for my ssh key when it starts up again?

- August 3: ```INFO Pulling debug logs from the bootstrap machine ************************************************************
ERROR failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain```

Which setup step could I be missing? (edited) 

- July 25: hey folks! I'm trying to setup a libvirt install, I've thus far gotten for the wait for bootstrap step, and failed on this
```INFO Waiting up to 30m0s for bootstrapping to complete... 
cdku    INFO Pulling debug logs from the bootstrap machine 
ERROR failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: dial tcp: lookup test1-rnrgr-bootstrap on 127.0.0.1:53:
 no such host 
FATAL failed to wait for bootstrapping to complete: timed out waiting for the condition```
did I mess up a step in this doc (https://github.com/openshift/installer/blob/master/docs/dev/libvirt/README.md#install-and-enable-libvirt)/am I missing something?
- July 22: @installer, is an SSH key required to install a cluster in 4.2, or have I hit a weird bug?

abhinavdahiya commented:
I think we should provide info that bootstrapping has failed due to reason. now we are gathering the logs

wking commented:
> I think we should provide info that bootstrapping has failed due to reason. now we are gathering the logs

Don't we do that now?  The FATAL error is why we gave up on the install (e.g. "never heard from the Kubernetes API").  The additional ERROR from log gathering is an error as well; I'd rather not drop it to INFO.  I'd be fine adding a prefix:

```go
logrus.Error(errors.Wrap(err2, "gathering bootstrap logs after install failure"))
```

which would result in output like:

```
ERROR gathering bootstrap logs after install failure: failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
FATAL waiting for Kubernetes API: context deadline exceeded
```

Do we think something like that would help with the misinterpretation?

patrickdillon commented:
> I think we should provide info that bootstrapping has failed due to reason. now we are gathering the logs
> Don't we do that now?
Not exactly, I think @abhinavdahiya is making the point that the ordering of error messages is confusing. Right now the errors look something like this: 
```
INFO Waiting up to 30m0s for the Kubernetes API at https://...
INFO Pulling debug logs from the bootstrap machine 
ERROR failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain 
FATAL waiting for Kubernetes API: context deadline exceeded 
```
An improvement would be something like this:
```
INFO Waiting up to 30m0s for the Kubernetes API at https://...
FATAL Installation failed with: waiting for Kubernetes API: context deadline exceeded 
INFO Pulling debug logs from the bootstrap machine 
ERROR failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain 
```
I think this suggestion makes sense and will work on implementing it unless a better suggestion comes along.

abhinavdahiya commented:
> > I think we should provide info that bootstrapping has failed due to reason. now we are gathering the logs
> > Don't we do that now?
> 
> Not exactly, I think @abhinavdahiya is making the point that the ordering of error messages is confusing. Right now the errors look something like this:
> 
> ```
> INFO Waiting up to 30m0s for the Kubernetes API at https://...
> INFO Pulling debug logs from the bootstrap machine 
> ERROR failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain 
> FATAL waiting for Kubernetes API: context deadline exceeded 
> ```
> 
> An improvement would be something like this (incorporating both suggestions):
> 
> ```
> INFO Waiting up to 30m0s for the Kubernetes API at https://...
> FATAL Installation failed with: waiting for Kubernetes API: context deadline exceeded 
> INFO Pulling debug logs from the bootstrap machine 
> ERROR gathering bootstrap logs after install failure: failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
> ```
> 
> I think this suggestion makes sense and will work on implementing it unless a better suggestion comes along.

Not really, the last message should be the **most important** message. but people are still tripping on an `ERROR` before the `FATAL` one..

So i think the best thing would be a combination of @wking suggestion and what i was suggesting..

```console
INFO Waiting up to 30m0s for the Kubernetes API at https://...
ERROR Pulling logs for debugging failure to bootstrap
ERROR Gathering bootstrap logs after failure: failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
FATAL waiting for Kubernetes API: context deadline exceeded 
```

patrickdillon commented:
Updated. Should now be:
```
INFO Waiting up to 30m0s for the Kubernetes API at https://...
INFO Installation failure detected. Running gather bootstrap..
INFO Pulling debug logs from the bootstrap machine
ERROR gathering debug logs after install failure: failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
FATAL waiting for Kubernetes API: context deadline exceeded 
```

wking commented:
> ```
> INFO Installation failure detected. Running gather bootstrap...
> INFO Pulling debug logs from the bootstrap machine
> ```

Sounds like the same thing twice?

wking commented on cmd/openshift-install/create.go:
```diff
@@ -96,8 +96,9 @@ var (
 
 				err = waitForBootstrapComplete(ctx, config, rootOpts.dir)
 				if err != nil {
+					logrus.Info("Installation failure detected. Running gather bootstrap...")
```
Maybe `logrus.Error(err)` here instead?  Overlaps with the later fatal message, but makes it harder to skim past the important bit ;).

wking commented on cmd/openshift-install/create.go:
```diff
@@ -96,8 +96,9 @@ var (
+					logrus.Info("Installation failure detected. Running gather bootstrap...")
 					if err2 := runGatherBootstrapCmd(rootOpts.dir); err2 != nil {
-						logrus.Error(err2)
+						logrus.Error(errors.Wrap(err2, "gathering debug logs after install failure"))
```
Alternatively, to really spell it out:

```go
logrus.Error(errors.Wrap(err2, "secondary error while gathering debug logs after fatal install failure"))
```

patrickdillon commented:
Ok, updated again, let's give this another try:

- moved the INFO message in gather, so it should only be displayed if ssh client is created
- added INFO/ERROR prefixes to maintain consistent capitalization

Error messages should now look like this...

**When SSH client fails**
```
INFO Waiting up to 30m0s for the Kubernetes API at https://...
ERROR Installation failure detected. Running gather bootstrap...
ERROR Unable to gather debug logs: failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
FATAL Bootstrap failed to complete: waiting for Kubernetes API: context deadline exceeded 
```

**When gather succeeds**
```
INFO Waiting up to 30m0s for the Kubernetes API at https://...
ERROR Installation failure detected. Running gather bootstrap...
INFO Pulling debug logs from the bootstrap machine
INFO Bootstrap gather logs captured here /home/padillon/myfavoriteplace
FATAL Bootstrap failed to complete: waiting for Kubernetes API: context deadline exceeded 
```

wking commented on cmd/openshift-install/create.go:
```diff
@@ -96,10 +96,11 @@ var (
 
 				err = waitForBootstrapComplete(ctx, config, rootOpts.dir)
 				if err != nil {
+					logrus.Error("Installation failure detected. Running gather bootstrap...")
```
I don't like `logrus.Error` for "here's our plan".  I'm fine with two info-level logs inside `runGatherBootstrapCmd` if you feel a need to separate SSH config from the SSH connection.  And I'm fine with an error-level log here that includes `err`.  And I'm fine with nothing here, because if the coming gather succeeds you won't have two separate errors for users to be confused about.

patrickdillon replied:
> I don't like logrus.Error for "here's our plan". 
> And I'm fine with nothing here, because if the coming gather succeeds you won't have two separate errors for users to be confused about.
It sounds like you have changed your mind about this https://github.com/openshift/installer/pull/2244#discussion_r315978679 or perhaps I am misinterpreting? To me this current implementation seems inline with ttps://github.com/openshift/installer/pull/2244#issuecomment-523242454

I don't feel like there is a clear direction here or that we are all going to reach an agreement, which is fine because I think this is subjective and difficult. I will revert this change and try to make a more minor change which will hopefully have at least some improvement.

wking replied:
> It sounds like you have changed your mind about this...

I'm still fine with an error-level log here as long as it includes `err`.  But I'm fine with 2992564e0 too.

patrickdillon commented:
Switched to a more minor approach:

```
INFO Waiting up to 30m0s for the Kubernetes API at https://...
INFO Pulling debug logs from the bootstrap machine
ERROR Attempted to gather debug logs after installation failure: failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
FATAL Bootstrap failed to complete: waiting for Kubernetes API: context deadline exceeded 
```

mtnbikenc commented:
/test e2e-aws-scaleup-rhel7
